README
# The CFT Home-Brew 16-bit Mini-Computer

The CFT is a home-designed, home-built mini-computer made from
scratch.

I've designed and built the instruction set, data path, processor,
computer, software stack, cross-assembler, emulator, software
toolchain and even some of the tools needed to do all this. You can
call this a ‘fantasy mini’.

This project has been going since approximately 2001, in my Copious
Free Time.


## Audience

There isn't one, so I couldn't tell you what it's like. This project
has been online for since 2005 and it's been noticed by four or five
people. Not quite enough for statistics.

**This is a personal project, and the repository is made public as a
curiosity. Expect everything to be wrong, broken, or dangerous.**

## Why?

Because it's fun.

To see how it's done.

To recall Those Heady Times.

To prove to myself I can do this.


## What?

The CFT is a long-term project built in my Copious Free Time. (a term used, as
always, sarcastically, hence the name) It's a homebrew, self-designed,
self-built Sixties-style 16-bit mini computer built out of 74-family integrated
circuits. Check out the
[ChangeLog](https://github.com/alexiosc/cft/blob/master/Changelog) for the
current status.

There may possibly be more stuff to look at, at https://www.bedroomlan.org/cft.

The project consists of:

* The CFT processor.
* Peripherals and cards, some implementing extensions to the CFT instruction set.
* A front panel with blinkenlights.
* A case for the lot.
* Documentation for everything, including the website and this.
* A cross-toolchain for working on the computer from a PC.

### The CFT Processor

* A 4 MHz clock. (250ns processor cycle)
* 16-bit word length. No byte access.
* 24-bit memory address bus.
* 16-bit I/O bus shared with the address bus. (the lower 1,024 addresses are easier to use)
* All instructions are exactly one 16-bit word long.
* Some instructions follow the PDP-8 ‘microcoded’ format with bitfields cotrolling combinations of actions. Bitwise negation, two's complement negation and incrementation are the same instruction with slightly different options.
* A relatively simple instruction set inspired by the PDP-8. This has moved on
  to something slightly more reminiscent of a modern processor.
* Microcoded design.
* Accumulator based architecture.
* Four user-accessible 16-bit registers with increment and decrement. Only the
  Accumulator (AC) is general purpose.
* PDP-8 style magic addresses working like auto-increment or auto-decrement
  index registers, or stack pointers.

### Peripherals

I originally sat down with Eagle CAD and went *crazy* with it. I had 50 sheets
of schematics for boards, all with a three-letter designator:

* MEM: A memory board with 512 KWords of RAM and 512 KWords of ROM.
* TTY: A quadruple serial board including both DCE and DTE
  connections, and a TTL one for an FTDI USB adaptor.
* INT: 8-interrupt controller board.
* IDE: Twin IDE bus interface for up to four drives.
* RTC: A real-time card with some NVRAM and three or six programmable timers.
* ETH: An offloaded Ethernet interface using the WIZ812 chip.
* KBD: A PS/2 Keyboard interface in pure discrete logic.
* PSG: A sound board with one or two AY-3-8910 chips and a SpeakJet® speech synthesiser.
* PFP: A Programmer's Front Panel capable of debugging microcode as well as full machine code.
* DEB: A DEBugging card to help diagnose problems with the hardware
  using a remote controller. (this was basically a simple, slow, specialised logic analyser)
* FDC: A Floppy Drive Controller for up to four drives, including 8-inch 50-pin ones.
* GIO: General Purpose I/O board including some digital I/O, two
  Atari-style digital joystick ports, a unidirectional parallel
  printer port, four user-controllable LEDs on the front panel (the
  PFP already had 16) and (on a whim) a cassette tape interface.
* MBU: processor extension to allow 21-bit memory addressing (up to 2 MWords).
* µCB: processor extension to allow multiple versions of microcode to co-exist. (I had spare space in the Microcode ROMs)
* VDU: a graphics card that does text at 80×25 and 40×25, and graphics
  at combinations of 160, 320, 640 columns and 100, 200, 400 rows with
  64 colours.

The currently planned peripherals are much more modest in scope, but
they also morphed quite a bit. The processor can now do 24-bit
addressing natively, the current microcode uses all of the available
ROM space and a bunch of cards have joined forces. The current state
of affairs looks like this:

* MEM: (tentative design) 512 kWords or 1 MWord of RAM, 512 KWords of
  ROM. Multiple cards can be installed. If density allows, there'll be
  more chip sockets for RAM.
* INT: 8 interrupt controller board.
* DFP: a big card containing both a remote debugging system and front panel drivers.
* TTY: serial port card, possibly colocated with…
* IDE: one or two IDE sockets for two or four IDE drives. (or CF
  cards, but I like the possibility of having some clunky old drive)
* SND: MIDI sound output. Possibly collocated with…
* VDU: a graphics card that does text at 80×30 and 40×30, and graphics
  at combinations of 160, 320, 640 columns and 120, 240, 480 rows
  using 64 colours. The new VDU card fits inside a CPLD and also
  includes a PS/2 interface.
* RTC: Real-time clock, timers.

I've built the MEM and DFP boards to pre-2019 specs and the computer was
working on them.

The 2019 redesign goes for proper PCBs for everything, so things are
taking a little extra time.

### Front Panel

I was really proud of the front panel before the 2019 redesign, it
worked great and showed the computer's entire status. There's support
for up to 160 lights, 31 switches and two keylocks for that PDP-8
feel.

The downside was the wiring: there were no fewer than *five* 50-wire
ribbon cables going to this crazy bastard.

The 2019 redesign keeps the layout, but can no longer display as much
of the computer's status on the front panel. I'm working on it. On the
upside, it now serialises data transmission and only a single 40-pin
cable is needed.

To cut costs, the original front panel was made out of eight modules,
each with up to 20 LEDs. The new front panel is made out of four
modules, each with up to 40 LEDs. Each module daisy-chains to the
next.

Switches are mounted on a separate plate and wired to a break-out
board, or the closest front panel modules on the 2019 version.


### The Case

Two words: wood paneling. Well, and a black acrylic front piece. The
rest will depend on how big the card cage gets, but I'll scratch make
the case out of sheet aluminium and wood.


### Documentation

There's just too much of it. The documentation was initially in LaTeX
and got up to 500 pages before I gave that up and started converting
it to HTML.

The 2019 redesign will need a redo of most of this, sadly, but the
current documents for hardware that's currently around are available
at https://www.bedroomlan.org/hardware/cft/book/.


### Software and Hardware Tools

I had to make my own tools for some of this, and *wanted* to make my
own for some other parts.

Hardware tools include
[Flashprog](https://www.bedroomlan.org/hardware/flashprog/), a
zero-parts flash programmer with a modern serial interface. The
interface formed the basis of the DEB and DFP cards' remote debugging
interface and I'm quite happy with it.

Software tools include:

* The [Microcode
  Assembler](https://www.bedroomlan.org/projects/mcasm/), also
  available on Github as https://github.com/alexiosc/mcasm.
* The CFT Assembler, written purely in Python and roughly based on
  PDP-8 Assemblers, but with some modern extras like namespaces and
  simple macros.
* Too many emulators to speak of!
    ** A basic microcode-level emulator in Python to verify the original
	design. This one is dead and buried, but still in the repository.
	** A Verilog emulator used to verify the hardware.
	** A C emulator that's fast enough to run software on.
	** A [Javascript microcode-level
    emulator](https://www.bedroomlan.org/hardware/cft/microcode-emulator/)
    that runs on the web. It's very slow, but it shows how the front
    panel works.
	** A newer JS-based emulator meant to me much faster. Still
    largely in the works, because I prefer the C-based emulator for
    testing.
* A testing framework that can run test suites on any emulator we can
  use.
* A basic Operating System for the CFT.
* A self-documenting Forth interpreter written in Assembly and Forth.


Changelog
2011-01-25  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Earliest draft version of the ‘memory mapped stack
	machine’ CPU description referencing a PDP-8-like design, but with an 8
	bit word.

2011-02-11  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Earliest version of the ‘simple 16-bit CPU’ description,
	a direct descendant of the Fungus CPU. Still references 18-bit
	registers (like Fungus).

2011-03-03  AC  <alexios@bedroomlan.org>

	[Tools] Created Subversion repository. Initial import of
	pre-existing code and data.

	[Des/Hardware] Removed Page Register from CPU design.

	[Des/Hardware] Changed MAR to '193-based register.

2012-03-12  AC  <alexios@bedroomlan.org>

	[Doc/Web] Created Facebook page.

2011-03-14  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Major changes to support autoincrement mode, fix
	issues with the L flag (excessive toggling), rework microcode,
	uaddr vector, control vector, register types, etc.

	[Des/Software/Microcode] Autoincrement now supported in LOAD.

	[Des/Hardware] Playing around with Eagle schematics as a very
	early feasibility study.

2011-08-08  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Many changes to the hardware design.

2011-08-30  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Switched to ROM-based ALU.

2011-08-31  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Changes to PCBs for new hardware style.

2011-09-02  AC  <alexios@bedroomlan.org>

	[Des/Hardware] More changes to the hardware and PCBs.

2011-10-24  AC  <alexios@bedroomlan.org>

	[Tools] Released ROMtools, a Python library to generate complex
	function tables on multiple ROMs. This is used to generate the ALU
	tables.

2011-10-25  AC  <alexios@bedroomlan.org>

	[Tools] Released mcasm (the Microcode Assembler).

2011-11-08  AC  <alexios@bedroomlan.org>

	[Tools] Released FlashProg hardware and software design.

2011-11-30  AC  <alexios@bedroomlan.org>

	[Tools] Removed VHDL support and relevant directories. The project
	has been using Verilog for verification for a while.

	[Misc] Created CFT Project logo.

2011-12-01  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] Worked on Front Panel schematics.

2011-12-03  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] First draft of the front panel.

2011-12-13  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] First release candidate of the front panel.

2011-12-19  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] Further design and verification work on the
	panel.

2011-12-21  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] More work on the front panel and its
	documentation.

2011-12-22  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP] Additional work on the front panel.

2011-12-26  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP/Controller] Completed first revision of Front
	Panel Controller board.

2011-12-28  AC  <alexios@bedroomlan.org>

	[Code/Cross-Assembler] Added macro and file inclusion support.

	[Tools] Slight changes to Eagle3D board rendering options.

	[Des/Hardware/PFP/Controller] Rendered front panel controller.

2011-12-30  AC  <alexios@bedroomlan.org>

	[Tools] Cleaned up crud entries from Subversion repository.

2012-01-09  AC  <alexios@bedroomlan.org>

	[Tools] Added functionality to extract bills of materials from
	Eagle schematics.

	[Doc/Book] Added BOM for front panel to documentation.

	[Doc/Extra] Added external documents useful for the development of
	the project (datasheets, etc).

	[Code/Software/Filesystem] First filesystem experiments.

	[Des/Hardware/PFP] Front Panel version 2 (rev B).

	[Doc/Hardware/PFP] Front Panel Bill of Materials generated.

2012-01-16  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP/Controller] SMD test of front panel
	controller. Superseded design, now using new version of front
	panel controller.

	[Des/Hardware/PFP] Various changes and corrections to the front
	panel. itself.

	[Doc/Gallery] Added PNG files to the gallery.

2012-01-20  AC  <alexios@bedroomlan.org>

	[Des/Hardware/PFP/Controller] Considerable amount of work on the
	front panel controller.

	[Des/Hardware] Worked on the clock generator.

2012-02-04  AC  <alexios@bedroomlan.org>

	[Des/Hardware/CTN] Drafting timer layout.

2012-02-14  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Corrected issues with the ROM-based ALU.

	[Testing,Des/Hardware] Coded ALU test benches in Verilog.

2012-02-17  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Fixed ALU bugs. Recoded Verilog ROM-ALU to handle
	the new ALU design.

	[Des/Hardware] Updated schematics, fixing numerous issues
	(including the ALU one).

	[Testing,Des/Hardware] Added batch tests for the ALU.

	[Doc/Extra] Added Forth documentation (external to the project).

	[Doc/Extra] Added FlashForth distribution (also external).

	[Code/Software/Filesystem] Work on the filesystem code.

2012-02-25  AC  <alexios@bedroomlan.org>

	[Des/Hardware] Completed CFT schematic collection into a very
	large Eagle file. This will help keep signal names constant,
	and will eventually aid DRC.

2012-03-09  AC  <alexios@bedroomlan.org>

	[Tools] Cleaning up Subversion repository. Removing temporary files.

2012-03-09  AC  <alexios@bedroomlan.org>

	[Code/Software/ROM/Forth] Coded core ROM Forth words (and then some).

2012-03-12  AC  <alexios@bedroomlan.org>

	[Code/Software/ROM/Forth] More Forth primitives. Code reorganisation.

	[Code/Software/ROM] First OS traps added.

2012-03-18  AC  <alexios@bedroomlan.org>

	[Testing,Des/Hardware] Added more hardware tests.

	[Des/Hardware] Imported mkskiptable, meant to speed up minor
	operation instructions. The functionality is currently too
	complex, so not used.

	[Code/Software/ROM] Bug fixes and Forth work.

2012-03-22  AC  <alexios@bedroomlan.org>

	[Code/Software/ROM/Forth] Implemented algorithms and unit tests
	for unsigned, symmetric and floored division in 16 and 32
	bits. For now, Forth algorithm tests are shared with the Verilog
	tests (and executed with the Verilog simulator).


asm/dfp.asm
;;; -*- cftasm -*-
;;; Definitions for the all new DFP (Debugging Front Panel) subsystem.
;;;

;;; Instructions (globally available)
	
.equ SOR	OUT dfp.DFP &00	        ; SOR: set Output Register
.equ ENEF	OUT dfp.DFP &08	        ; ENEF: enable feature
.equ DISEF	OUT dfp.DFP &09	        ; DISEF: disable feature
.equ SENTINEL	OUT dfp.DFP &0f	        ; SENTINEL instruction
.equ PRINTA     OUT dfp.DFP &10	        ; Print out an address
.equ PRINTC     OUT dfp.TX	        ; Print out a character
.equ PRINTD     OUT dfp.DFP &12	        ; Print AC in base 10, signed
.equ PRINTU     OUT dfp.DFP &13	        ; Print AC in base 10, unsigned
.equ PRINTH     OUT dfp.DFP &14	        ; Print AC in base 16
.equ PRINTB     OUT dfp.DFP &15	        ; Print AC in base 2
.equ PRINTSP    OUT dfp.DFP &16	        ; Print out a space
.equ PRINTNL    OUT dfp.DFP &17	        ; Print out a newline
.equ DEBUGON    OUT dfp.DFP &18	        ; Start tracing (slow)
.equ DEBUGOFF   OUT dfp.DFP &19	        ; Stop tracing (slow)
.equ DUMP       OUT dfp.DFP &1a	        ; Dump current state
.equ PRINTHI    OUT dfp.DFP &1b	        ; Set a HI word for PRINTLO
.equ PRINTLO    OUT dfp.DFP &1c	        ; Print out HI:LO words
.equ HALT	OUT dfp.DFP &1d	        ; Halt
.equ SUCCESS    OUT dfp.DFP &1e	        ; Assert: true
.equ FAIL       OUT dfp.DFP &1f	        ; Assert: false

.equ LSR        IN dfp.DFP &00	        ; LSR: load Switch Register
.equ LDSR       IN dfp.DFP &01	        ; LDSR: load DIP switch register
.equ QEF        IN dfp.DFP &08	        ; QEF: query features (see below)

.pushns dfp

.equ DFP R &100				; Base address

;;; I/O Registers 
	
.equ RX   DFP &11		        ; Last read byte from DFP console
.equ TX   DFP &11		        ; Write to the DFP console
.equ ICR  DFP &0a		        ; ICR: interrupt control register
.equ ISR  DFP &0a		        ; ISR: interrupt status register

;;; Flags for the QEF instruction

.equ QEF_PFP  #--------'-------1        ; PFP: full front panel installed
.equ QEF_DEB  #--------'------1-        ; DEB functionality present
.equ QEF_TTY  #--------'-----1--        ; Debug TTY present
.equ QEF_TST  #--------'----1---        ; Testing interface (PRINTx) present
.equ QEF_TRC  #--------'---1----        ; Tracing (DEBUGON/OFF) available
.equ QEF_UTR  #--------'--1-----	; Microcode tracing available
.equ QEF_DMP  #--------'-1------	; DUMP available
.equ QEF_SNT  #--------'1-------	; SENTINEL available
.equ QEF_HOF  #-------1'--------	; Will halt on FAIL
.equ QEF_HOS  #------1-'--------	; Will halt on SENTINEL
.equ QEF_LCK  #-----1--'--------	; Panel is locked
.equ QEF_DVE  #000-----'--------	; Detection/version: verilog
.equ QEF_DHW  #010-----'--------	; Detection/version: hardware
.equ QEF_DCE  #110-----'--------        ; Detection/version: C-emulated panel
.equ QEF_DJE  #111-----'--------        ; Detection/version: JS-emulated panel
.equ QEF_DET  #111-----'--------        ; Detection mask

;;; Flags for the ENEF/DISEF instruction
		
.equ FTR_HOB  #---------------1         ; Halt on bus errors (emulator only)
.equ FTR_TRC  #-----------1----		; Assembly trace
.equ FTR_UTR  #----------1-----		; Microcode trace
.equ FTR_HOS  #------1---------		; Halt on SENTINEL

;;; Flags for the ISR register

.equ ISR_IRQ6 #---------------1         ; IRQ6 asserted
.equ ISR_IRQ1 #--------------1-         ; IRQ1 asserted (IFR1 switch operated)
.equ ISR_IFR6 #-------------1--         ; IFR6 switch operated
.equ ISR_TTYQ #------------1---         ; TTY characters available

;;; Flags for the ICR register
		
.equ ICR_IFR1 #--------------1-         ; Signal IRQ1 on IFR1 press
.equ ICR_IFR6 #-------------1--         ; Signal IRQ6 (or IRQ) on IFR6 press
.equ ICR_TTYQ #------------1---         ; IRQ6 when TTY characters available

;;; Macro: enef(lit)
;;;
;;; Enable features. Flags are given as literals, using the FTR_xxx values
;;; above.
;;;
;;; Side effects:
;;;   AC = lit
;;;   DFP features enabled.
.macro enef (lit)
		LI %lit			; enef(%lit)
		ENEF
.end
					; TTY characters available

;;; Macro: disef(lit)
;;;
;;; Disable features. Flags are given as literals, using the FTR_xxx values
;;; above.
;;;
;;; Side effects:
;;;   AC = lit
;;;   DFP features disabled.
.macro disef (lit)
		LI %lit			; disef(%lit)
		DISEF
.end
		


.popns
	
;;; End of file.
